cmd/influx/write.go

```diff
@@ -68,94 +62,133 @@ or add an entire file specified with an @ prefix.`
+			if err != nil {
+				return fmt.Errorf("failed to decode bucket-id: %v", err)
+			}
+		}
+		if writeFlags.Bucket != "" {
+			filter.Name = &writeFlags.Bucket
 		}
-	}
-	if writeFlags.Org != "" {
-		filter.Org = &writeFlags.Org
-	}
 
-	buckets, n, err := bs.FindBuckets(ctx, filter)
-	if err != nil {
-		return fmt.Errorf("failed to retrieve buckets: %v", err)
-	}
+		if writeFlags.org.id != "" {
+			filter.OrganizationID, err = platform.IDFromString(writeFlags.org.id)
+			if err != nil {
+				return fmt.Errorf("failed to decode org-id id: %v", err)
+			}
+		}
+		if writeFlags.org.name != "" {
+			filter.Org = &writeFlags.org.name
+		}
 
-	if n == 0 {
-		if writeFlags.Bucket != "" {
-			return fmt.Errorf("bucket %q was not found", writeFlags.Bucket)
+		buckets, n, err := bs.FindBuckets(ctx, filter)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve buckets: %v", err)
 		}
 
-		if writeFlags.BucketID != "" {
-			return fmt.Errorf("bucket with id %q does not exist", writeFlags.BucketID)
+		if n == 0 {
+			if writeFlags.Bucket != "" {
+				return fmt.Errorf("bucket %q was not found", writeFlags.Bucket)
+			}
+
+			if writeFlags.BucketID != "" {
+				return fmt.Errorf("bucket with id %q does not exist", writeFlags.BucketID)
+			}
 		}
-	}
 
-	bucketID, orgID := buckets[0].ID, buckets[0].OrgID
+		bucketID, orgID = buckets[0].ID, buckets[0].OrgID
+	}
 
 	var r io.Reader
-	if args[0] == "-" {
-		r = os.Stdin
-	} else if len(args[0]) > 0 && args[0][0] == '@' {
-		f, err := os.Open(args[0][1:])
+	if len(args) > 0 && args[0][0] == '@' {
+		// backward compatibility: @ in arg denotes a file
+		writeFlags.File = args[0][1:]
+	}
+
+	if len(writeFlags.File) > 0 {
+		f, err := os.Open(writeFlags.File)
 		if err != nil {
-			return fmt.Errorf("failed to open %q: %v", args[0][1:], err)
+			return fmt.Errorf("failed to open %q: %v", writeFlags.File, err)
 		}
 		defer f.Close()
 		r = f
+		if len(writeFlags.Format) == 0 && strings.HasSuffix(writeFlags.File, ".csv") {
+			writeFlags.Format = inputFormatCsv
+		}
+	} else if len(args) == 0 || args[0] == "-" {
+		// backward compatibility: "-" also means stdin
+		r = os.Stdin
 	} else {
 		r = strings.NewReader(args[0])
 	}
+	// validate input format
+	if len(writeFlags.Format) > 0 && writeFlags.Format != inputFormatLineProtocol && writeFlags.Format != inputFormatCsv {
+		return fmt.Errorf("unsupported input format: %s", writeFlags.Format)
+	}
 
-	s := write.Batcher{
-		Service: &http.WriteService{
-			Addr:               flags.Host,
-			Token:              flags.Token,
-			Precision:          writeFlags.Precision,
-			InsecureSkipVerify: flags.skipVerify,
-		},
+	if writeFlags.Format == inputFormatCsv {
+		r = write.CsvToProtocolLines(r)
 	}
 
-	ctx = signals.WithStandardSignals(ctx)
-	if err := s.Write(ctx, orgID, bucketID, r); err != nil && err != context.Canceled {
-		return fmt.Errorf("failed to write data: %v", err)
+	if writeFlags.DryRun {
```

Comment:
curiosity: feels like the Dry Run functionality can be pulled out from this large func. It shares some things, but those shared things you can pull out as needed and clean up the rest of this here. This command was already doing a fair amount, the dry run stuff baked into here makes it harder to follow. WDYT?

Reply:
At the time when I was initially writing this, I thought that it is less error-prone to have a single function that is executed for both write and 'write with dry-run'. Now, after the changes, when the flags settle down and doing a few steps back to see the code better, I have to agree that the code complexity will reduce if there will be separate functions for each command that will call a shared function to obtain a line data reader. Good point, thx :)